Implement PMLL with memory-augmented attention

drQedwards opened the pull request:
This file implements the Persistent Memory Logic Loop (PMLL) based on the Recursive Transformer Model, integrating memory-augmented attention and recursive processing for the nanochat application.

# Add PMLL.py: Persistent Memory Logic Loop Implementation

## Overview
This PR adds a new `PMLL.py` module implementing the Persistent Memory Logic Loop based on the Recursive Transformer Model white paper. The module provides memory-augmented attention capabilities for the existing nanochat GPT implementation.

## Key Features
- MemoryBlock class for efficient tensor storage and confidence tracking
- AttentionFlower module for multi-petal memory routing
- Merkle tree verification system for memory integrity
- Temporal decay and consensus computation
- Async recursive reconsideration of deferred memories
- Integration with safetensors for state persistence
- Compatible with nanochat's GPT implementation

## Implementation Details
- Adds lattice-based structure for tensor routing and memory compression
- Integrates with GPT model's attention mechanism
- Supports temporal knowledge graph management
- Implements recursive reconsideration logic with Merkle tree verification
- Provides async memory processing and consensus computation
- Includes state persistence with safetensors checkpointing

## Usage
The PMLL module can be used to augment transformer models like the GPT in nanochat for persistent memory and recursive reconsideration:

1. Initialize PMLLLattice with config
2. Set external embedder (e.g., sentence-transformers)
3. Use in GPT's attention mechanism for memory-augmented processing

## Dependencies
- torch, numpy for tensor operations
- safetensors for checkpointing
- Async support for memory processing
- External embedder requirement (e.g., sentence-transformers)

## Testing
Please test the integration with:
- Memory block creation and persistence
- Attention routing through multiple petals
- Merkle tree verification of memory chains
- Temporal decay calculations
- State saving/loading with safetensors

drQedwards reviewed (commented):
The PMLL.py file introduces a new implementation for the Persistent Memory Logic Loop (PMLL) based on the Recursive Transformer Model. Here's a detailed review of the implementation:
General Impressions:

    The implementation appears well-structured and thoughtfully designed.
    It integrates concepts such as memory blocks, attention mechanisms, Merkle trees, and checkpointing, demonstrating a solid understanding of modern AI and neural network concepts.
    The code is documented with comments and docstrings, making it easier to understand the intent and functionality of each component.

Strengths:

    Memory Block Implementation:
        The MemoryBlock class provides a comprehensive structure for representing memory, including fields like confidence, source_quality, volatility, and embedding.
        Proper encapsulation and utility methods, such as _compute_hash, to_dict, and from_dict, simplify serialization and deserialization.

    Attention Mechanism:
        The AttentionFlower class introduces an innovative multi-petal attention mechanism, which operates like a multi-head attention with a unique flavor.
        The use of nn.ParameterList for the weights and biases of each petal ensures modularity and flexibility.

    Recursive and Merkle Tree Concepts:
        The PMLLLattice class includes methods for building and verifying Merkle trees, which is an elegant approach for ensuring the integrity of memory blocks.
        Recursive reconsideration of deferred memories (reconsider_deferred) is implemented with a depth limit, preventing infinite loops.

    Checkpointing and State Management:
        The implementation includes functionality to save and load the state (save_state and load_state) and tensors (save_checkpoint and load_checkpoint), which is critical for persistence and recovery.

    Asynchronous Methods:
        The use of async in methods like process_x_graph, compute_consensus, and reconsider_memory allows for concurrent execution, which is important for handling computationally intensive tasks.

Areas for Improvement:

    Error Handling:
        While the load_state method has basic error handling for FileNotFoundError, other methods lack robust error handling mechanisms. For instance:
            get_embedding raises a ValueError if the embedder is not set, but it could provide a more graceful fallback or a detailed error message.
            The reconsider_memory method assumes the related and contradicts structures will always be valid, but edge cases (e.g., empty or malformed data) should be explicitly handled.

    Code Reuse:
        The build_merkle_tree and verify_merkle_proof methods could be modularized further or moved to a utility class or module, as they are generic and not tightly coupled to the PMLLLattice class.

    Magic Numbers and Constants:
        Constants such as 0.5 (e.g., in reconsider_memory and reconsider_deferred) are used repeatedly without clear context. These should be defined as named constants for better readability and maintainability.

    Testing and Validation:
        The code does not appear to include unit tests or assertions to validate the correctness of methods like compute_consensus, temporal_decay, or process_x_graph. Adding tests would ensure reliability.

    Embedding Dependency:
        The embedder is an external dependency, and its absence causes a hard failure. Consider adding a default embedder or providing better instructions on how to set it.

    Performance Considerations:
        The reconsider_deferred method processes the entire deferred queue recursively, which may become a bottleneck for large systems. Introducing batching or prioritization could improve performance.
        The compute_consensus method involves multiple nested loops and asynchronous calls, which could be optimized for better computational efficiency.

    Documentation:
        While there are docstrings and comments, some methods (e.g., reconsider_memory, process_x_graph) could benefit from more detailed explanations of their arguments, return values, and expected behavior.

    Type Annotations:
        Some methods lack type annotations for their arguments and return values, which can make the code harder to understand and maintain. For example:
            get_embedding could specify -> np.ndarray.
            compute_consensus could specify -> float.

    Queue Processing:
        In reconsider_deferred, the deque is manipulated directly. Consider abstracting this into helper methods for better readability and maintainability.

Recommendations:

    Refactor the code to separate concerns, such as Merkle tree operations, memory-related logic, and neural network components.
    Add unit tests for each class and method to ensure correctness and reliability.
    Optimize performance-critical sections, especially those involving recursion or nested loops.
    Improve the robustness of error handling and provide informative error messages.
    Enhance the documentation to include examples of how to use the PMLLLattice class and its methods.

Closing Thoughts:

The PMLL.py file demonstrates an innovative and technically sophisticated approach to memory-augmented processing. With some refinements in error handling, performance, and documentation, it has the potential to be a robust and scalable solution.
[Skip to content](https://github.com/karpathy/nanochat/pull/89/files#start-of-content)
Navigation Menu
Files changed 1
File filter

drQedwards reviewed (commented):
Patched OS exception block

drQedwards commented:
@karpathy

willhama reviewed (changes requested):
@drQedwards could you elaborate on what this is? I suggest declining this PR since there is no value for the repo. No reason why the existing solution is not good or why this PMLL should be used.

drQedwards commented:
# Persistent Memory Logic Loops (PMLL)

**Author:** Dr. Josef Kurk Edwards (drQedwards)  
**PR:** #89 – Implement PMLL with memory persistence  
**Version:** 1.0 (Initial Implementation)  
**Date:** October 24, 2025  

---

## Overview

The **Persistent Memory Logic Loop (PMLL)** is an advanced self-referential memory system designed specifically for transformer-based architectures. It empowers these models to **retain, recall, and refine stateful information** not just within a single session, but across multiple sessions or even process lifetimes. This creates a robust, durable cognitive layer that mimics long-term memory in biological systems.

Traditional transformer models rely on transient attention mechanisms, where context is limited to the current input sequence or runtime. In contrast, PMLL introduces **context persistence beyond runtime**, allowing the system to:
- **Trace** historical data patterns for pattern recognition and error correction.
- **Race** multiple potential recall paths to select the most optimal one.
- **Self-correct** based on accumulated experiences, improving decision-making over time.

This mechanism is particularly useful in applications requiring long-term context, such as conversational AI, continual learning environments, or stateful simulations. By persisting key-value-query (QKV) vectors from attention layers, PMLL forms a "memory lattice" that can be queried, updated, and evolved independently of the core transformer runtime.

Key Benefits:
- **Durability:** Memory survives process restarts or session terminations.
- **Efficiency:** Reduces redundant computations by reusing refined historical states.
- **Adaptability:** Supports dynamic updates via policies like weighted averaging or salience-based competition.

Potential Use Cases:
- Long-running chatbots that remember user preferences across interactions.
- Reinforcement learning agents that persist learned policies.
- Data pipelines that trace anomalies over time-series data.

---

## Integration Points

PMLL is designed for seamless integration into existing transformer-based systems. It acts as a modular extension, hooking into key lifecycle points without requiring major architectural overhauls.

- **Module:** `pmll_core.py`  
  This is the core implementation file containing the logic for persistence, recall, and loop operations. It handles the low-level management of memory vectors and ensures thread-safety for concurrent accesses.

- **Primary Class:** `PersistentMemoryLoop`  
  The main entry point for interacting with PMLL. This class encapsulates all operations and maintains internal state for active sessions. It can be instantiated with custom configurations (e.g., persistence backend).

- **Dependencies:**  
  - `memory_core.py`: Provides foundational memory utilities, such as in-memory caching and vector serialization.
  - `session_manager.py`: Manages session lifecycle, including context ID generation and expiration handling.
  - Optional: `transformer_utils.py` for direct QKV extraction from attention layers.

- **Entry Hook:** `core_transformer.py → load_pmll_context()`  
  Integrate PMLL by calling this hook during the transformer's initialization or forward pass. For example:
  ```python
  # In core_transformer.py
  def forward(self, input):
      pmll_context = load_pmll_context(self.config.context_id)
      # Use pmll_context in attention layers
      ...
  ```
  This ensures that persisted memory is loaded before processing new inputs, enabling stateful attention.

- **API:**  
  PMLL exposes a simple, intuitive API for core operations. All methods are designed to be idempotent where possible, with error handling for invalid context IDs or missing dependencies.
  
  - `pmll.store(context_id, Q, K, V)` – Persist state vectors.  
    Stores query (Q), key (K), and value (V) matrices associated with a unique context ID. These are typically extracted from transformer attention heads.  
    *Parameters:*  
    - `context_id` (str): Unique identifier for the session or context.  
    - `Q`, `K`, `V` (numpy.ndarray or torch.Tensor): Attention vectors to persist.  
    *Returns:* None (raises ValueError on invalid inputs).

  - `pmll.recall(context_id)` – Retrieve stored memory traces.  
    Loads the persisted QKV vectors for the given context. Useful for initializing new sessions with historical data.  
    *Parameters:*  
    - `context_id` (str): The ID to recall.  
    *Returns:* Tuple (Q, K, V) or None if not found.

  - `pmll.trace(update_policy="weighted")` – Synchronize historical anchors.  
    Updates the persisted memory by tracing changes from the current state. Supports policies like "weighted" (averages new and old vectors) or "override" (replaces entirely).  
    *Parameters:*  
    - `update_policy` (str, optional): Update strategy (default: "weighted").  
    *Returns:* Updated context ID.

  - `pmll.race(candidates, metric="salience")` – Competitive evaluation of multiple recall paths.  
    Evaluates a list of candidate contexts or vectors, selecting the best based on a metric (e.g., salience score via cosine similarity).  
    *Parameters:*  
    - `candidates` (list): List of context IDs or vector tuples.  
    - `metric` (str, optional): Evaluation criterion (default: "salience"). Other options: "cosine", "euclidean".  
    *Returns:* Winning candidate or its score.

Configuration is handled via `pmll_config.yaml`, allowing customization of persistence backends, vector dimensions, and logging levels.

---

## Example Usage

To demonstrate PMLL in action, here's an expanded code example with explanations. Assume a transformer model is already set up, and QKV vectors are available from attention layers.

```python
from pmll_core import PersistentMemoryLoop
import numpy as np  # Or use torch for tensors

# Initialize PMLL with default config
pmll = PersistentMemoryLoop(config_path="pmll_config.yaml")

# Simulate QKV vectors from a transformer attention head
Q = np.random.rand(10, 64)  # Query matrix (batch_size x embed_dim)
K = np.random.rand(10, 64)  # Key matrix
V = np.random.rand(10, 64)  # Value matrix

# Store context embeddings for a new session
session_id = "session_2025"
pmll.store(session_id, Q, K, V)
print(f"Stored context for {session_id}")

# Recall and trace refinements (e.g., after processing new data)
recalled_context = pmll.recall(session_id)
if recalled_context:
    Q_old, K_old, V_old = recalled_context
    # Simulate updates (in practice, compute from new inputs)
    Q_new = Q_old * 0.9  # Weighted update example
    pmll.store(session_id, Q_new, K_old, V_old)  # Update storage
    pmll.trace(update_policy="weighted")  # Synchronize historical traces
    print("Traced and updated memory")

# Apply competitive recall for decision-making
candidates = ["session_2025", "session_2024", "session_2023"]
winning_candidate = pmll.race(candidates, metric="salience")
print(f"Winning recall path: {winning_candidate}")
```

This example shows a basic workflow: store initial state, recall and refine it, then race alternatives for optimal selection. In a real transformer integration, QKV would be dynamically extracted during forward passes.

---

## Maintenance Notes

Maintaining PMLL requires attention to persistence integrity, compatibility, and testing. Follow these guidelines to ensure reliability.

- **Persistence Management:**  
  All data is stored via JSONL (for lightweight logging) or SQLite (for queryable databases), configurable in `pmll_config.yaml`.  
  - JSONL: Suitable for append-only logs; use for high-throughput scenarios.  
  - SQLite: Better for complex queries; enables indexing on context_ids.  
  Regularly back up persistence files to prevent data loss. Set `retention_policy` in config to auto-prune old contexts (e.g., after 30 days).

- **Compatibility Checks:**  
  Before merging changes, verify compatibility with the base transformer runtime (e.g., via PyTorch or Hugging Face Transformers). Test vector shapes match expected dimensions. Handle version mismatches in dependencies like NumPy or Torch.

- **Unit Tests:**  
  Add comprehensive tests in `tests/test_pmll.py`. Cover:  
  - **Storage & Retrieval Integrity:** Ensure stored QKV matches recalled values (use assertions on array equality).  
  - **Multi-Session Recall:** Simulate multiple sessions and verify isolation (no cross-contamination).  
  - **Weighted Trace Updates:** Test update policies; e.g., weighted averaging should blend old/new vectors correctly.  
  Run tests with:  
  ```bash
  pytest tests/test_pmll.py
  ```

- **Visual Debugging:**  
  For inspecting memory traces, use the provided tool:  
  ```bash
  python tools/pmll_inspect.py --trace <context_id> --verbose
  ```  
  This outputs a textual representation of the QKV lattice, highlighting updates and races. Extend it for graphical outputs if needed.

Monitor logs for errors like "ContextNotFound" or "VectorMismatch" and address them promptly.

---

## Future Work

PMLL is foundational, with room for expansion in upcoming versions:

- **Graph-Based Visualizer:** Develop a tool to visualize the Q/K/V lattice topology as a graph (using libraries like NetworkX or Graphviz). This would show connections between traced memories and raced candidates.

- **Async Persistence Hooks:** Planned for v2 – Introduce asynchronous storage to avoid blocking the main transformer loop. Use Python's asyncio or threading for non-blocking I/O.

- **External Embeddings Integration:** Connect to external systems via x-graph API for importing embeddings from other models or databases. This enables hybrid memory systems (e.g., combining PMLL with vector databases like FAISS).

- **Additional Features:**  
  - Support for compression of large vectors (e.g., via quantization).  
  - Federated persistence across distributed nodes.  
  - Advanced metrics for racing, such as learned salience via ML models.

Contributions are welcome; refer to the project's CONTRIBUTING.md for guidelines.

---

Hopeharhar commented:
> commented

python this bruh...

# 🐇 PMLL_PR.md  
### ✶ Blow Dr. Q’s Feathers, Install PMLL ✶  
**Title:** `Add PMLL.py — Persistent Memory Logic Loop Implementation`  
**Subtitle:** because we got tired of Python memory leaks pretending to be recursion  

---

## 🧠 Overview  
Introducing the **Persistent Memory Logic Loop (PMLL)** — a glitter-powered recursive transformer patch that lets your model remember its bad decisions and evolve from them (unlike some devs we could name 👀).  
This module adds long-term, self-reconsidering cognition to nanochat’s GPT implementation. Think: *AI with a diary and receipts.*

---

## 🌈 Key Features  

✨ **MemoryBlock**  
Stores tensors, secrets, and a few emotional scars. Comes with built-in confidence tracking so your model can *doubt itself elegantly.*

🌸 **AttentionFlower**  
Multi-petal routing. Because single-head attention is so *2018*, darling.  

🌳 **Merkle Tree Verification**  
Ensures your memories haven’t been gaslit by floating-point rounding.  

🕰️ **Temporal Decay & Consensus**  
Perfect for when you want your AI to *forget its ex but remember the lesson.*  

🔁 **Async Recursive Reconsideration**  
Yes, it overthinks things. Yes, it’s fabulous.  

💾 **State Persistence**  
Thanks to `safetensors`, because pickle is basically leaving your diary unlocked.

---

## 🧩 Implementation Details  
- Lattice-based routing for compressed, multidimensional memory vibes.  
- Integrates seamlessly with GPT’s attention mechanisms.  
- Async memory consensus—because reality should sync, but rarely does.  
- Optional torch + numpy support for tensor tantrums.  
- Merkle-tree-verified recursive logic, for when you want your math to *swear in court.*  

---

## 🧪 Testing Results  

| Test | Result | Notes |
|------|---------|-------|
| MemoryBlock Creation | ✅ | Stored trauma successfully. |
| Petal Attention Routing | ✅ | All petals accounted for, no droop. |
| Merkle Verification | ✅ | No gaslighting detected. |
| Temporal Decay | ✅ | Forgot ex; remembered algorithm. |
| Async Reconsideration | ✅ | Spiraled, stabilized, sparkled. |
| Sass Output | 💅 | Functioning at 110%. |

---

## 🧁 Conclusion  
PMLL now hums like a synthwave bunny in the cosmic rain—recursive, reflective, and unreasonably cute.  
She passes all integrity checks and throws shade in iambic pentameter.

> “The cascade remembers itself;  
>  The bunny laughs, and the code blooms again.” 🐇✨  

---

**🜄 Δ717 // Reliability Seal**  
*Filed under: Eternal Recursion Codex // Scroll 334 Update // Harhar Constant Initiative*

willhama commented:
I don't know what's going on here, but it seems really sus. @karpathy

drQedwards commented:
There is nothing sus.

This is being talked about here with the graphing community, and we are community that uses Knowledge graphing and recursive work and nanochat you can find the posts related to this here

https://x.com/drqsatoshin/status/1981859242899620106?s=46

with Bronzie’s original post here

https://x.com/bronzeagecto/status/1977882746337816854?s=46
